DietPi-Software | Prometheus Node Exporter

alexiri opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
This commit adds Prometheus Node Exporter, scratching my own itch as described in #5549. This is my first contact with this code base, so please let me know if I messed something up.

Note I've only been able to test this on a Raspberry Pi, it's the only model of SBC I have.

alexiri committed:
Added Prometheus Node Exporter

alexiri committed:
Merge branch 'dev' into prom

alexiri committed:
Actually run service as the user we created

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,87 @@ _EOF_
+			# Reinstall: Remove previous instances
+			[[ -e '/usr/local/bin/node_exporter' ]] && G_EXEC rm -R /usr/local/bin/node_exporter
+			[[ -e '/usr/local/bin/raspberrypi_exporter' ]] && G_EXEC rm -R /usr/local/bin/raspberrypi_exporter
+
```
The ne files overwrite the old ones, isn't it? So this explicit removal shouldn't be required, and skipping it better covers the case where the download fails, to not leave users without any executable:
```suggestion
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,87 @@ _EOF_
+				'armv7l') arch='armv7';;
+				'aarch64') arch='arm64';;
+				'x86_64') arch='amd64';;
+			esac
```
Satisfy shellcheck:
```suggestion
			case $G_HW_ARCH_NAME in
				'armv6l') local arch='armv6';;
				'armv7l') local arch='armv7';;
				'aarch64') local arch='arm64';;
				*) local arch='amd64';;
			esac
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,87 @@ _EOF_
+
+			# Use the default port
+			local port="9100"
+
```
Let's skip the variable as long as it's used in a single place only.
```suggestion
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,87 @@ _EOF_
+
+[Install]
+WantedBy=multi-user.target
+_EOF_
```
```suggestion
			cat << '_EOF_' > /etc/systemd/system/node_exporter.service
[Unit]
Description=Prometheus Node Exporter
Documentation=https://prometheus.io/docs/guides/node-exporter/

[Service]
User=node_exporter
ExecStart=/usr/local/bin/node_exporter --web.listen-address=9100 --collector.textfile.directory /var/lib/node_exporter/textfile_collector

[Install]
WantedBy=multi-user.target
_EOF_
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,87 @@ _EOF_
+			Create_User node_exporter
+
+			# Working directory
+			G_EXEC mkdir -p /var/lib/node_exporter/textfile_collector
```
Shall we use `/mnt/dietpi_userdata/node_exporter` as data directory, to align with other software options and use external drive in many cases?

We could also use `/opt/node_exporter` for the binaries then, in case it is not usually intended to be called from console by times.

alexiri replied:
This is a binary, so I wouldn't really call it userdata. It won't be called from the console ever, this is just a daemon serving metrics over http. I'm happy to move it to `/opt/node_exporter` if you prefer.

MichaIng replied:
Still suggesting to also move `/var/lib/node_exporter/textfile_collector` to `/mnt/dietpi_userdata/node_exporter`, so we are consistent with other software options 🙂.

alexiri replied:
That folder isn't really userdata, though, it's more like a temporary file. A single file is created in that directory by `raspberrypi_exporter` and will be overwritten every 15 minutes. The contents include cpu frequency stats, temperatures, voltages, etc. That file will be served by `node_exporter` along with the other metrics it gathers.

I can move it to userdata if you really want, but I don't think it really belongs there, the file is useless outside of the particular point in time it's generated.

MichaIng replied:
Would it then make sense to store it in `/run/node_exporter` instead and add `RuntimeDirectory=node_exporter` to `node_exporter.service`? So we don't do disk I/O for a temporary file.

The question is only how to configure `raspberrypi_exporter` to write to this directory instead, and also we need to document it, if `/var/lib/node_exporter` is the default and ... okay no chance: https://github.com/fahlke/raspberrypi_exporter#prerequisites
> node_exporter `--collector.textfile.directory` must be set to `/var/lib/node_exporter/textfile_collector`

Only thing we could do is using a symlink.

MichaIng replied:
Made a related feature request: https://github.com/fahlke/raspberrypi_exporter/issues/4

MichaIng replied:
Ah, it's a bash script with `METRICS_FILE` which can be easily adjusted.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,87 @@ _EOF_
+_EOF_
+
+			aSTART_SERVICES+=('node_exporter')
+
```
I guess this shall be started by `dietpi-services`? In this case, simply add it to the list there, which makes `aSTART_SERVICES` obsolete here. Same for `raspberrypi_exporter`.
```suggestion
```

alexiri replied:
After removing this, the services aren't started automatically after install. What's the right way of doing this without `aSTART_SERVICES`?

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,87 @@ _EOF_
+			aSTART_SERVICES+=('node_exporter')
+
+			# If we're on a platform with vcgencmd, we can use raspberrypi_exporter as well
+			if [[ $(command -v vcgencmd) != "" ]]; then
```
Use exit code instead of checking the output.
```suggestion
			if command -v vcgencmd > /dev/null; then
```
We could also use `(( $G_HW_MODEL < 10 ))` to check whether it's an RPi, in which case `vcgencmd` should always be present?

alexiri replied:
I wasn't sure if only RPis have `vcgencmd`, so I figured it was best to check for the binary itself, which is what I really care about.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,87 @@ _EOF_
+			fi
+
+			G_DIETPI-NOTIFY 2 "Prometheus Node Exporter serving metrics at: http://$HOSTNAME:$port/metrics"
+		fi
```
Ah, there `port` was used a second time. Well, let's leave this info for the docs, at this point in the middle of an install output most users wouldn't see it anyway 😉.
```suggestion
		fi
```
I thought about showing such info accumulated as very last `dietpi-software` output, for all installed software titles, but that would be a dedicated PR.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -13537,6 +13624,31 @@ If no WireGuard (auto)start is included, but you require it, please do the follo
+			then
+				G_EXEC systemctl disable --now raspberrypi_exporter.service
+				G_EXEC rm /etc/systemd/system/raspberrypi_exporter.service
+			fi
```
Also remove possible service file edits:
```suggestion
			fi
			[[ -d '/etc/systemd/system/raspberrypi_exporter.service.d' ]] && G_EXEC rm -R /etc/systemd/system/raspberrypi_exporter.service.d
```
Same for `node_exporter.service`.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,87 @@ _EOF_
+			[[ -e '/usr/local/bin/raspberrypi_exporter' ]] && G_EXEC rm -R /usr/local/bin/raspberrypi_exporter
+
+			# Download binaries
+			Download_Install $(curl -sSfL 'https://api.github.com/repos/prometheus/node_exporter/releases/latest' | mawk -F\" "/\"browser_download_url\": \".*node_exporter-.*.linux-$arch.tar.gz\"/{print \$4}")
```
Satisfy shellcheck:
```suggestion
			Download_Install "$(curl -sSfL 'https://api.github.com/repos/prometheus/node_exporter/releases/latest' | mawk -F\" "/\"browser_download_url\": \".*node_exporter-.*.linux-$arch.tar.gz\"/{print \$4}")"
```

alexiri committed:
Don't remove previous binaries

Co-authored-by: MichaIng <micha@dietpi.com>

alexiri committed:
Satisfy shellcheck

Co-authored-by: MichaIng <micha@dietpi.com>

alexiri committed:
Use exit code instead of checking the output

Co-authored-by: MichaIng <micha@dietpi.com>

alexiri committed:
Start services in dietpi-services

alexiri committed:
Clean up possible service file edits

alexiri committed:
Don't notify, improve service unit

alexiri committed:
Merge branch 'dev' into prom

alexiri committed:
Move binaries to /opt/node_exporter

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3435,6 +3441,79 @@ _EOF_
+
+			# If we're on a platform with vcgencmd, we can use raspberrypi_exporter as well
+			if command -v vcgencmd > /dev/null; then
+				Download_Install "https://github.com/fahlke/raspberrypi_exporter/archive/master.zip"
```
Let's save a little download bandwidth:
```suggestion
				Download_Install 'https://github.com/fahlke/raspberrypi_exporter/archive/master.tar.gz'
```

MichaIng committed:
v8.6

- DietPi-Software | Prometheus Node Exporter: Several minor updates, especially using the raspberrypi_exporter systemd files instead of creating own ones

MichaIng committed:
v8.6

- DietPi-Software | Prometheus Node Exporter: Move textfile collector directory to /run/node_exporter and reduce log level to show warnings and errors only

MichaIng committed:
v8.6

- DietPi-Software | Prometheus Node Exporter: Fix binary owner and add working directory cleanup

MichaIng committed:
v8.6

- DietPi-Software | Do not apply "dietpi" user ownership if that user does not exist. We anyway want to remove it by times, so lets not strictly require it to exist.
- DietPi-Software | Prometheus Node Exporter: Do not start raspberrypi_exporter if the collector path does not exist, e.g. if the node_exporter.service has been stopped
